keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,958 @@
+
+### Non-Goals
+
+1. "Fix" openapi v3 to openapi v2 conversion
+  This is a non-goal for two reasons:
+    * These formats are not compatible, and there WILL be data loss and inaccuracy
+    * This negates the benefits of using OpenAPI v3 for the richer type information
+2. Provide general-purpose OpenAPI visualization.
+
+
+## Proposal
+
+
+### Basic Usage
+
+The following user experience should be possible with `kubectl explain`
+
+```shell
+kubectl explain pods.spec
+```
+
+Output should be familiar to users of today's `kubectl explain`, except new
+information from the OpenAPI v3 spec is now populated.
+
+Note: Feature during development will be gated by an experimental flag. The commands
+shown here elide the experimental flag for clarity.
+
+### Built-in Template Options
+#### Plaintext
+
+
+```shell
+kubectl explain pods
+```
+or
+
+```shell
+kubectl explain pods --output plaintext
+```
+
+The plaintext output format is the default and should be crafted to be as close
+as the existing `explain` output in use before this KEP.
+
+#### OpenAPIV3 (raw json)
+
+```shell
+kubectl explain pods --output openapiv3
+```
+
+To get raw OpenAPI v3 data for a certain resource today involves:
+1.) setting up kubectl proxy
+2.) fetching the correct path at `/openapi/v3/<group>/<version>`
+3.) filtering out unwanted results
+
+This command is useful not only for its convenience, but also other visualizations
+may be built upon the raw output if we opt not to support a first-class custom
+template solution in the future.
+
+#### HTML
+
+```shell
+kubectl explain pods --output html
+```
+
+Similarly to [godoc](https://pkg.go.dev), the we suggest to provide a searchable,
+navigable, generated webpage for the kubernetes types of whatever cluster kubectl
+is talking to.
+
+#### Markdown
+
+```shell
+kubectl explain pods --output md
+```
+
+When using the `md` template, a markdown document is printed to stdout, so it
+might be saved and used for a documentation website, for example.
+
+### Risks and Mitigations
+
+#### OpenAPI V3 Not Available
+
+##### Risk
+
+OpenAPI v3 data is not available in the current cluster.
+
+##### Mitigation
+
+###### If the user does not provide an --output argument
+
+This is a "compatibility mode" where users will see OpenAPI v3 plaintext template
+if the data is available, or fallback to old `kubectl explain` behavior for openapi v2.
+
+If the initial request for OpenAPI V3 fails, the old implementation using
+OpenAPI v2 should be used instead. There is another risk that this fallback process
+takes too long for interactive speed, so timeout should be carefully considered
+to prevent user from waiting too long to see their results.
+
+###### If the user does provide an --output argument
+
+If a user specifies an `--output` argument and the server 404's attempting to
+fetch the correct openapi version for the template, a new error message should
+be thrown to the effect of: `server missing openapi data for version: %v.%v.%v`.
```

Comment:
I've been taking `--output` being used to mean that `--experimental-openapiv3` is assumed, at least for the alpha phase. After alpha `--experimental-openapiv3` is removed.

The fallback case only applies to those using `kubectl explain <fieldpath>` without any of the new `output` or `experimental-openapiv3` options. If those are specified, it is inferred that user wants openapiv3, or an error.

Reply:
Would `--experimental-output` fix all of that? As long as you use the experimental-output flag, you're OK with using OpenAPI v3, if you're aksing for plaintext it can fallback to v2, if not it will fail with an error?

Reply:
I like `--experimental-output`. Updating the kep.

However, if it is specified, I think not having a fallback is best. Users of this feature I think should have the expectation that the template rendering is being used, or no rendering at all.

Reply:
Sounds great to me.